openlibrary/plugins/upstream/borrow.py

```diff
@@ -129,6 +129,19 @@ def POST(self, key):
         if not edition:
             raise web.notfound()
 
+        from openlibrary.book_providers import get_book_provider
+
+        # Direct to the first web book if at least one is available.
+        if (
+            action == "borrow"
+            and (provider := get_book_provider(edition))
+            and (providers := provider.get_ebook_providers(edition))
+            and len(providers) >= 1
+            and providers[0].access == "open-access"
+        ):
```

Comment:
This is for action="borrow" which is typically for borrowable books, but given most trusted book providers are for open access materials (and we tend to use "read" for these) we should make sure a case of /borrow?action=read also succeeds